File: apina-core/src/main/java/fi/evident/apina/model/settings/TranslationSettings.java
package fi.evident.apina.model.settings;

import fi.evident.apina.model.type.ApiTypeName;
import fi.evident.apina.utils.PatternSet;

import java.util.Collection;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;

import static java.util.Collections.unmodifiableCollection;

/**
 * Various settings guiding the translation.
 */
public final class TranslationSettings {

    public final PatternSet blackBoxClasses = new PatternSet();
    private final TreeMap<String, ImportDefinition> importsByModule = new TreeMap<>();
    private final Set<ApiTypeName> importedTypes = new TreeSet<>();

    public boolean isBlackBoxClass(String name) {
        return blackBoxClasses.test(name);
    }

    public void addImport(String moduleName, Collection<String> types) {
        ImportDefinition importDefinition = importsByModule.computeIfAbsent(moduleName, ImportDefinition::new);

        for (String type : types) {
            ApiTypeName typeName = new ApiTypeName(type);
            if (!importedTypes.add(typeName))
                throw new IllegalArgumentException("type " + typeName + " is already imported");

            importDefinition.addType(typeName);
        }
    }

    public Collection<ImportDefinition> getImports() {
        return unmodifiableCollection(importsByModule.values());
    }

    public boolean isImported(ApiTypeName typeName) {
        return importedTypes.contains(typeName);
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/model/settings/TranslationSettings.kt
package fi.evident.apina.model.settings

import fi.evident.apina.model.type.ApiTypeName
import fi.evident.apina.spring.NameTranslator
import fi.evident.apina.utils.PatternSet
import java.util.*

/**
 * Various settings guiding the translation.
 */
class TranslationSettings {

    val blackBoxClasses = PatternSet()
    private val controllersToProcess = PatternSet()
    private val endpointUrlMethods = PatternSet()
    private val importsByModule = TreeMap<String, ImportDefinition>()
    private val importedTypes = TreeSet<ApiTypeName>()
    val brandedPrimitiveTypes = mutableListOf<BrandedPrimitiveType>()
    val nameTranslator = NameTranslator()
    var platform = Platform.ANGULAR

    @Deprecated("Writing types as interfaces is preferred and selection will be removed in future")
    var typeWriteMode = TypeWriteMode.INTERFACE
    var optionalTypeMode = OptionalTypeMode.NULL
    var enumMode = EnumMode.DEFAULT
    var removedUrlPrefix = ""
    var reexportImports = false

    fun isBlackBoxClass(name: String) = name in blackBoxClasses

    fun addImport(moduleName: String, types: Collection<String>) {
        val importDefinition = importsByModule.getOrPut(moduleName) { ImportDefinition(moduleName) }

        for (type in types) {
            val typeName = ApiTypeName(type)
            if (!importedTypes.add(typeName))
                throw IllegalArgumentException("type $typeName is already imported")

            importDefinition.addType(typeName)
        }
    }

    val imports: Collection<ImportDefinition>
        get() = importsByModule.values

    fun isImportedOrBrandedType(typeName: ApiTypeName) =
        typeName in importedTypes || brandedPrimitiveTypes.any { typeName == it.brandedType }

    fun isProcessableController(name: String) =
        controllersToProcess.isEmpty || name in controllersToProcess

    fun isUrlEndpoint(className: String, methodName: String) =
        "$className.$methodName" in endpointUrlMethods

    fun addControllerPattern(pattern: String) {
        controllersToProcess.addPattern(pattern)
    }

    fun addEndpointUrlMethodPattern(pattern: String) {
        endpointUrlMethods.addPattern(pattern)
    }

    fun normalizeUrl(url: String): String =
        url.removePrefix(removedUrlPrefix)

}
